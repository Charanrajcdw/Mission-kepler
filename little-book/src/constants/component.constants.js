export const MODAL = {
  members: "Members",
  newBlog: "Add New Blog",
  remove: "Remove",
};

export const MEMBERS_LIST = {
  title: "Members",
  noMembers: "No members available",
};

export const BLOG_FORM = {
  title: "Add New Blog",
  button: "ADD",
  titlePlaceholder: "Name your blog",
  detailsPlaceholder: "Write Content Here ..",
  invalidValues: "Invalid title or description",
  invalidBlog: "Blog title already exists",
  blogAdded: "New Blog added successfully",
  localBlog: "local",
};

export const WARNING_MODAL = {
  title: "Unsaved changes will be lost. Are you sure to continue?",
  continue: "Continue",
  cancel: "Cancel",
};
