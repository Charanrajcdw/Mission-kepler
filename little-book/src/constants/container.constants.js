export const SIDEBAR = {
  logo: "Little Book",
  filter: "FILTER",
  noFilter: "Add blogs to show filters",
  viewMembers: "View Members",
  switchToDark: "Switch to Dark Mode",
  switchToLight: "Switch to Light Mode",
  blogs: "Blogs",
};
