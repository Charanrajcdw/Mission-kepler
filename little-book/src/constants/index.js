export * from "./url.constants";
export * from "./theme.constants";
