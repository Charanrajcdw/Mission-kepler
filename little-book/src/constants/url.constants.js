export const API_URLS = {
  blogs: "https://jsonmockserver.vercel.app/api/blogs",
  users: "https://jsonmockserver.vercel.app/api/users",
  images: "https://jsonmockserver.vercel.app/",
};
