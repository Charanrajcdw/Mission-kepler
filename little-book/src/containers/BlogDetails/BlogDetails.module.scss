@import "../../scss/fonts";

.blogDetailsContainer {
  background-color: var(--primary-bg);
  width: 50.75%;
  padding: 3rem 4rem 3rem 3.75rem;
  height: 100vh;
  overflow: auto;

  &::-webkit-scrollbar {
    display: none;
  }

  .noBlog {
    font-size: 2rem;
    color: var(--primary-font-bg);
  }

  textarea {
    background-color: var(--primary-bg);
    color: var(--primary-font-bg);
    width: 100%;
    outline: none;
    resize: none;
    border: none;
  }

  .blogTitle {
    margin: 28px 0 16px;
    font-size: 2.2rem;
    font-weight: 500;
    font-family: $font-roboto;
  }

  .blogDetails {
    font-size: 0.95rem;
    line-height: 1.5;
    opacity: 0.8;
    &::-webkit-scrollbar {
      display: none;
    }
  }

  .buttonsContainer {
    display: flex;
    flex-flow: row nowrap;
    justify-content: flex-end;
    margin: 15px 0;
    gap: 15px;
  }
}
