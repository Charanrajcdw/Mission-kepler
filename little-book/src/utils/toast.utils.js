import { toast } from "react-toastify";

export const showSuccessToast = (message, currentTheme) => {
  toast.success(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};

export const showWarningToast = (message, currentTheme) => {
  toast.warn(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};
