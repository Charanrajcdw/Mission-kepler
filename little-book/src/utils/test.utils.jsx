import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import ThemeContextProvider from "../contexts/theme.context";
import blogSlice from "../store/slices/blogSlice";
import { THEME } from "../constants";

const customRender = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
    ...options
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return render(ui, { wrapper: ProviderWrapper, ...options });
};

const customCreateSnapshot = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return renderer.create(<ProviderWrapper>{ui}</ProviderWrapper>).toJSON();
};

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { customRender as render, customCreateSnapshot as createSnapshot };
