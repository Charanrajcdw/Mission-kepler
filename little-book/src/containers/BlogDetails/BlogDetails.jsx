import { useEffect, useContext, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import { ToastContainer } from "react-toastify";
import "react-toastify/dist/ReactToastify.min.css";
import styles from "./BlogDetails.module.scss";
import { Image, Button, Loader } from "../../components";
import { showSuccessToast, showWarningToast } from "../../utils/toast.utils";
import { ThemeContext } from "../../contexts/theme.context";
import { BLOG_DETAILS } from "../../constants";
import { blogActions } from "../../store";

const BlogDetails = () => {
  const blogTitleRef = useRef();
  const blogDetailsRef = useRef();
  const { currentTheme } = useContext(ThemeContext);
  const { blogData, currentBlog, isLoaded, isEditing } = useSelector((state) => state.blogs);
  const dispatch = useDispatch();

  useEffect(() => {
    blogTitleRef.current && heightHandler(blogTitleRef.current);
    blogDetailsRef.current && heightHandler(blogDetailsRef.current);
  }, [currentBlog]);

  // handler function for onClick on edit blog button
  const editHandler = () => {
    dispatch(blogActions.modifyEditStatus(true));
  };

  // handler function for onClick on save edit button
  const saveHandler = () => {
    const title = blogTitleRef.current.value.trim();
    const details = blogDetailsRef.current.value.trim();
    const filteredBlogData = blogData.filter((blog) => blog.title !== currentBlog.title);
    if (title.length <= 0 || details.length <= 0) {
      showWarningToast(BLOG_DETAILS.invalidValues, currentTheme);
    } else if (filteredBlogData.some((blog) => blog.title.toLowerCase() === title.toLowerCase())) {
      showWarningToast(BLOG_DETAILS.invalidBlog, currentTheme);
    } else {
      showSuccessToast(BLOG_DETAILS.saveSuccess, currentTheme);
      dispatch(blogActions.editExistingBlog({ title, details }));
    }
  };

  // handler function for onClick on cancel edit button
  const cancelHandler = () => {
    dispatch(blogActions.modifyEditStatus(false));
  };

  // handler function to scroll the blogs list to top
  const heightHandler = (element) => {
    element.style.height = "0px";
    element.style.height = element.scrollHeight + "px";
  };

  const textareaBuilder = (styleName, value, refValue) => {
    return (
      <textarea
        className={styleName}
        value={isEditing ? undefined : value}
        readOnly={!isEditing}
        onChange={(event) => heightHandler(event.target)}
        ref={refValue}
      />
    );
  };

  return (
    <article className={styles.blogDetailsContainer}>
      <ToastContainer />
      {isLoaded ? (
        !currentBlog ? (
          <p className={styles.noBlog}>{BLOG_DETAILS.noBlog}</p>
        ) : (
          <>
            <Image className="blogImg" src={currentBlog.photo} alt={currentBlog.title} />
            {textareaBuilder(styles.blogTitle, currentBlog.title, blogTitleRef)}
            {textareaBuilder(styles.blogDetails, currentBlog.details, blogDetailsRef)}
            <div className={styles.buttonsContainer}>
              {isEditing ? (
                <>
                  <Button className="blueBtn smallBtn" clickHandler={cancelHandler}>
                    {BLOG_DETAILS.cancelText}
                  </Button>
                  <Button className="pinkBtn smallBtn" clickHandler={saveHandler}>
                    {BLOG_DETAILS.saveText}
                  </Button>
                </>
              ) : (
                <Button className="blueBtn smallBtn" clickHandler={editHandler}>
                  {BLOG_DETAILS.editText}
                </Button>
              )}
            </div>
          </>
        )
      ) : (
        <Loader />
      )}
    </article>
  );
};

export default BlogDetails;
