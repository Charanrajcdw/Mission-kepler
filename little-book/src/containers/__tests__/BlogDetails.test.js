import { render, userEvent, screen, createSnapshot } from "../../utils/test.utils";
import BlogDetails from "../BlogDetails/BlogDetails";
import { BLOG_DETAILS } from "../../constants";

jest.mock("../../components/Loader/Loader", () => () => <div data-testid="loader" />);
const blogData = {
  title: "Blog title",
  details: "Blog details",
  photo: "https://cdn.mos.cms.futurecdn.net/E3JYf6eJHQawCL2AR5cBv4.jpg",
  type: "Local",
};
const blogsData = [
  { title: "title 1", type: "international" },
  { title: "title 2", type: "regional" },
];
const updateBlogData = {
  title: "New blog title",
  details: "New blog details",
  photo: "https://cdn.mos.cms.futurecdn.net/E3JYf6eJHQawCL2AR5cBv4.jpg",
  type: "Local",
};

describe("Blog details UI", () => {
  it("should render loader component if data is loading", () => {
    const initialBlogsData = { isLoaded: false };
    render(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    const loader = screen.getByTestId("loader");
    expect(loader).toBeInTheDocument();
  });

  it("should render no blogs content if there is no current blog", () => {
    const initialBlogsData = { isLoaded: true };
    render(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    const noBlogsElement = screen.getByText(BLOG_DETAILS.noBlog);
    expect(noBlogsElement).toBeInTheDocument();
  });

  it("should render blog content if the blog is loaded and not in editing state", () => {
    const initialBlogsData = { isLoaded: true, isEditing: false, currentBlog: blogData };
    render(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    const titleElement = screen.getByDisplayValue(blogData.title);
    const detailsElement = screen.getByDisplayValue(blogData.details);
    const editButton = screen.getByText(BLOG_DETAILS.editText);
    expect(titleElement).toBeInTheDocument();
    expect(detailsElement).toBeInTheDocument();
    expect(editButton).toBeInTheDocument();
  });

  it("should render blog content if the blog is loaded and in editing state", () => {
    const initialBlogsData = { isLoaded: true, isEditing: true, currentBlog: blogData };
    render(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    const cancelButton = screen.getByText(BLOG_DETAILS.cancelText);
    const saveButton = screen.getByText(BLOG_DETAILS.saveText);
    expect(cancelButton).toBeInTheDocument();
    expect(saveButton).toBeInTheDocument();
  });
});

describe("Blog details functionalities", () => {
  it("should have same data if cancel button is clicked", async () => {
    const user = userEvent.setup();
    const initialBlogsData = { isLoaded: true, isEditing: false, currentBlog: blogData };
    render(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    const titleElement = screen.getByDisplayValue(blogData.title);
    const detailsElement = screen.getByDisplayValue(blogData.details);
    const editButton = screen.getByText(BLOG_DETAILS.editText);
    await user.click(editButton);
    await user.type(titleElement, updateBlogData.title);
    await user.type(detailsElement, updateBlogData.details);
    const cancelButton = await screen.findByText(BLOG_DETAILS.cancelText);
    await user.click(cancelButton);
    expect(titleElement).toHaveValue(blogData.title);
    expect(detailsElement).toHaveValue(blogData.details);
  });

  it("should have updated data if save button is clicked", async () => {
    const user = userEvent.setup();
    const initialBlogsData = { isLoaded: true, isEditing: false, currentBlog: blogData, blogData: blogsData };
    render(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    const titleElement = screen.getByDisplayValue(blogData.title);
    const detailsElement = screen.getByDisplayValue(blogData.details);
    const editButton = screen.getByText(BLOG_DETAILS.editText);
    await user.click(editButton);
    await user.clear(titleElement);
    await user.type(titleElement, updateBlogData.title);
    await user.clear(detailsElement);
    await user.type(detailsElement, updateBlogData.details);
    const saveButton = await screen.findByText(BLOG_DETAILS.saveText);
    await user.click(saveButton);
    expect(titleElement).toHaveValue(updateBlogData.title);
    expect(detailsElement).toHaveValue(updateBlogData.details);
  });

  it("should show warning if same title is entered or empty values found", async () => {
    const user = userEvent.setup();
    const initialBlogsData = { isLoaded: true, isEditing: false, currentBlog: blogData, blogData: blogsData };
    render(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    const titleElement = screen.getByDisplayValue(blogData.title);
    const detailsElement = screen.getByDisplayValue(blogData.details);
    const editButton = screen.getByText(BLOG_DETAILS.editText);
    // testing for empty values
    await user.click(editButton);
    await user.clear(titleElement);
    await user.clear(detailsElement);
    const saveButton = await screen.findByText(BLOG_DETAILS.saveText);
    await user.click(saveButton);
    const valuesWarningElement = await screen.findByText(BLOG_DETAILS.invalidValues);
    expect(valuesWarningElement).toBeInTheDocument();
    // testing for same title
    await user.type(titleElement, blogsData[0].title);
    await user.type(detailsElement, updateBlogData.details);
    await user.click(saveButton);
    const titleWarningElement = await screen.findByText(BLOG_DETAILS.invalidBlog);
    expect(titleWarningElement).toBeInTheDocument();
  });
});

describe("Blog details snapshots", () => {
  it("should match blog details snapshot in display mode", () => {
    const initialBlogsData = { isLoaded: true, isEditing: false, currentBlog: blogData };
    const blogDetails = createSnapshot(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    expect(blogDetails).toMatchSnapshot();
  });

  it("should match blog details snapshot in editing mode", () => {
    const initialBlogsData = { isLoaded: true, isEditing: true, currentBlog: blogData };
    const blogDetails = createSnapshot(<BlogDetails />, { preloadedState: { blogs: initialBlogsData } });
    expect(blogDetails).toMatchSnapshot();
  });
});
