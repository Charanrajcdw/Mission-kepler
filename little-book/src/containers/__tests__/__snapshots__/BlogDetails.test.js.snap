// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog details snapshots should match blog details snapshot in display mode 1`] = `
<article
  className="blogDetailsContainer"
>
  <div
    className="Toastify"
  />
  <div
    className="blogImg"
  >
    <img
      alt="Blog title"
      className="img"
      onError={[Function]}
      src="https://cdn.mos.cms.futurecdn.net/E3JYf6eJHQawCL2AR5cBv4.jpg"
    />
  </div>
  <textarea
    className="blogTitle"
    onChange={[Function]}
    readOnly={true}
    value="Blog title"
  />
  <textarea
    className="blogDetails"
    onChange={[Function]}
    readOnly={true}
    value="Blog details"
  />
  <div
    className="buttonsContainer"
  >
    <button
      className="btn blueBtn smallBtn"
      onClick={[Function]}
    >
      EDIT CONTENT
    </button>
  </div>
</article>
`;

exports[`Blog details snapshots should match blog details snapshot in editing mode 1`] = `
<article
  className="blogDetailsContainer"
>
  <div
    className="Toastify"
  />
  <div
    className="blogImg"
  >
    <img
      alt="Blog title"
      className="img"
      onError={[Function]}
      src="https://cdn.mos.cms.futurecdn.net/E3JYf6eJHQawCL2AR5cBv4.jpg"
    />
  </div>
  <textarea
    className="blogTitle"
    onChange={[Function]}
    readOnly={false}
  />
  <textarea
    className="blogDetails"
    onChange={[Function]}
    readOnly={false}
  />
  <div
    className="buttonsContainer"
  >
    <button
      className="btn blueBtn smallBtn"
      onClick={[Function]}
    >
      CANCEL
    </button>
    <button
      className="btn pinkBtn smallBtn"
      onClick={[Function]}
    >
      SAVE CONTENT
    </button>
  </div>
</article>
`;
